source/SH3/system/glcontext.cpp

```diff
@@ -17,11 +17,109 @@
 #include "SH3/system/glcontext.hpp"
 #include "SH3/system/log.hpp"
 #include "SH3/system/window.hpp"
+#include "SH3/glcallback.hpp"
 #include <GL/glew.h>
 #include <GL/glu.h>
 #include <GL/glext.h>
 #include <glm/gtc/matrix_transform.hpp>
 
+/**
+ *  OpenGL error callback function. Used in conjunction with glDebugMessageCallback
+ *
+ *  @arg source     The Source of this error
+ *  @arg type       The type of error this is
+ *  @arg id         Error ID
+ *  @arg severity   How severe this error is. If it's fatal, terminate the program with @ref die()
+ *  @arg length     Length of the message log
+ *  @arg message    Error message passed to us from OpenGL
+ *  @arg data       User Param data
+ */
+__glCallback debugCallback(GLenum source, GLenum type, GLuint id, GLenum severity, GLsizei length, const GLchar *message, const void *data)
+{   
+    LogLevel    level;
+    std::string errMsg;
+    
+    errMsg = "";
+
+    // Get the source of the error message
+    switch(source)
+    {
+    case GL_DEBUG_SOURCE_API: // Error caused by the OpenGL API
+        errMsg += "API ";
+        break;
+    case GL_DEBUG_SOURCE_WINDOW_SYSTEM:
+        errMsg += "WINDOW SYSTEM ";
+        break;
+    case GL_DEBUG_SOURCE_SHADER_COMPILER:
+        errMsg += "SHADER COMPILER ";
+        break;
+    case GL_DEBUG_SOURCE_APPLICATION:
+        errMsg += "APPLICATION ";
+        break;
+    case GL_DEBUG_SOURCE_OTHER:
+        errMsg += "OTHER ";
+        break;
+    default:
+        errMsg += "UNKNOWN ";
+        break;
+    }
+
+    // Get the type of error this is
+    switch(type)
+    {
+    case GL_DEBUG_TYPE_ERROR:
+        errMsg += "(ERROR): ";
+        break;
+    case GL_DEBUG_TYPE_DEPRECATED_BEHAVIOR:
+        errMsg += "(DEPRECATED): ";
+        break;
+    case GL_DEBUG_TYPE_UNDEFINED_BEHAVIOR:
+        errMsg += "(UNDEFINED BEHAVIOR): ";
+        break;
+    case GL_DEBUG_TYPE_PORTABILITY:
+        errMsg += "PORTABILITY! ";
+        break;
+    case GL_DEBUG_TYPE_PERFORMANCE:
+        errMsg += "(PERFORMANCE): ";
+        break;
+    case GL_DEBUG_TYPE_OTHER:
+        errMsg += "(OTHER): ";
+        break;
+    case GL_DEBUG_TYPE_MARKER:
+        errMsg += "(MARKER): ";
+        break;
+    default:
+        errMsg += "(UNKNOWN): ";
+        break;
+    }
+
+    switch(severity)
+    {
+        case GL_DEBUG_SEVERITY_HIGH:
+            level = LogLevel::FATAL;
+            break;
+        case GL_DEBUG_SEVERITY_MEDIUM:
+            level = LogLevel::ERROR;
+            break;
+        case GL_DEBUG_SEVERITY_LOW:
+            level = LogLevel::WARN;
+            break;
+        case GL_DEBUG_SEVERITY_NOTIFICATION:
+            level = LogLevel::INFO;
+            break;
+        default:
+            level = LogLevel::INFO;
+            break;
+    }
+
+    errMsg += message;
+
+    if(severity == GL_DEBUG_SEVERITY_HIGH)
```

Comment:
I like always having braces. I think we've also always used braces on statements elsewhere.

Reply:
Yeah same here, I wasn't sure if I should do it for consistency. I think I need for fix up a few files, though if use a formatter it should sort itself out.